jobfunnel/backend/scrapers/base.py

```diff
@@ -0,0 +1,449 @@
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from multiprocessing import Lock, Manager
+from time import sleep, time
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+from bs4 import BeautifulSoup
+from requests import Session
+from requests.adapters import HTTPAdapter
+from tqdm import tqdm
+from urllib3.util import Retry
+
+from jobfunnel.backend import Job, JobStatus
+from jobfunnel.backend.tools import Logger
+from jobfunnel.backend.tools.delay import calculate_delays
+from jobfunnel.backend.tools.filters import JobFilter
+from jobfunnel.resources import (MAX_CPU_WORKERS, USER_AGENT_LIST, JobField,
+                                 Locale)
+
+if False:  # or typing.TYPE_CHECKING  if python3.5.3+
+    from jobfunnel.config import JobFunnelConfigManager
+
+
+class BaseScraper(ABC, Logger):
+    """Base scraper object, for scraping and filtering Jobs from a provider
+    """
+
+    def __init__(self, session: Session, config: 'JobFunnelConfigManager',
+                 job_filter: JobFilter) -> None:
+        """Init
+
+        Args:
+            session (Session): session object used to make post and get requests
+            config (JobFunnelConfigManager): config containing all needed paths,
+                search proxy, delaying and other metadata.
+            job_filter (JobFilter): object for filtering incoming jobs using
+                various internal filters, including a content-matching tool.
+                NOTE: this runs-on-the-fly as well, and preempts un-promising
+                job scrapes to minimize session() usage.
+
+        Raises:
+            ValueError: if no Locale is configured in the JobFunnelConfigManager
+        """
+        # Inits
+        super().__init__(
+            level=config.log_level,
+            file_path=config.log_file,
+        )
+        self.job_filter=job_filter
+        self.session=session
+        self.config=config
+        if self.headers:
+            self.session.headers.update(self.headers)
+
+        # Elongate the retries TODO: make configurable
+        retry = Retry(connect=3, backoff_factor=0.5)
+        adapter = HTTPAdapter(max_retries=retry)
+        self.session.mount('http://', adapter)
+        self.session.mount('https://', adapter)
+
+        # Ensure that the locale we want to use matches the locale that the
+        # scraper was written to scrape in:
+        if self.config.search_config.locale != self.locale:
+            raise ValueError(
+                f"Attempting to use scraper designed for {self.locale.name} "
+                "when config indicates user is searching with "
+                f"{self.config.search_config.locale.name}"
+            )
+
+        # Ensure our properties satisfy constraints
+        self._validate_get_set()
+        self.thread_manager = Manager()
+
+        # Construct actions list which respects priority for scraping Jobs
+        self._actions_list = [(True, f) for f in self.job_get_fields]
```

Comment:
This is the line of code where the debugger starts to stall out.

Reply:
ah very interesting, I haven't used PyCharm myself, but it would seem that it is not correctly handling the abstract inheritance of the properties.